IdentityCore/src/webview/embeddedWebview/MSIDOAuth2WebviewController.m

```diff
@@ -26,12 +26,185 @@
+    if ( _completionHandler )
+    {
+        void (^completionHandler)(MSIDWebOAuth2Response *response, NSError *error) = _completionHandler;
+        _completionHandler = nil;
+        
+        //TODO: think about how to generate different type of response from URL
+        MSIDWebOAuth2Response *response = [MSIDWebOAuth2Response new];
+        dispatch_async( dispatch_get_main_queue(), ^{
+            completionHandler(response, error);
+        });
+    }
+    
+    [_completionLock unlock];
+}
+
+#pragma mark - MSIDWebviewDelegate
+
+- (void)webAuthDidStartLoad:(NSURL*)url
+{
+}
+
+- (void)webAuthDidFinishLoad:(NSURL*)url
+{
+}
+
+- (BOOL)webAuthShouldStartLoadRequest:(NSURLRequest *)request
+{
+    NSString *requestURL = [request.URL absoluteString];
+    
+    // Stop at the end URL.
+    if ([[requestURL lowercaseString] hasPrefix:[_webviewRequest.endURL.absoluteString lowercaseString]])
+    {
+        NSURL* url = request.URL;
+        [self webAuthDidCompleteWithURL:url];
+        
+        // Tell the web view that this URL should not be loaded.
+        return NO;
+    }
+    
+    return YES;
+}
+
+// The user cancelled authentication
+- (void)webAuthDidCancel
+{
+    // Dispatch the completion block
+    NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorUserCancel, @"The user has cancelled the authorization.", nil, nil, nil, nil, nil);
+    [self endWebAuthenticationWithError:error orURL:nil];
+}
+
+// Authentication completed at the end URL
+- (void)webAuthDidCompleteWithURL:(NSURL *)endURL
+{
+    [self endWebAuthenticationWithError:nil orURL:endURL];
+}
+
+// Authentication failed somewhere
+- (void)webAuthDidFailWithError:(NSError *)error
+{
+    // Ignore WebKitError 102 for OAuth 2.0 flow.
+    if ([error.domain isEqualToString:@"WebKitErrorDomain"] && error.code == 102)
+    {
+        return;
+    }
+    
+    // Prior to iOS 10 the WebView trapped out this error code and didn't pass it along to us
+    // now we have to trap it out ourselves.
+    if ([error.domain isEqualToString:NSCocoaErrorDomain] && error.code == NSUserCancelledError)
+    {
+        return;
+    }
+    
+    // If we failed on an invalid URL check to see if it matches our end URL
+    if ([error.domain isEqualToString:@"NSURLErrorDomain"] && (error.code == -1002 || error.code == -1003))
+    {
+        NSURL* url = [error.userInfo objectForKey:NSURLErrorFailingURLErrorKey];
+        NSString* urlString = [url absoluteString];
+        if ([[urlString lowercaseString] hasPrefix:_webviewRequest.endURL.absoluteString.lowercaseString])
+        {
+            [self webAuthDidCompleteWithURL:url];
+            return;
+        }
+        
+    }
+    
+    if (NSURLErrorCancelled == error.code)
+    {
+        //This is a common error that webview generates and could be ignored.
+        //See this thread for details: https://discussions.apple.com/thread/1727260
+        return;
+    }
+    
+    if([error.domain isEqual:@"WebKitErrorDomain"])
```

Comment:
I checked, WebKitErrorDomain includes the following errors:

`WebKitErrorCannotShowMIMEType =                             100,
WebKitErrorCannotShowURL =                                  101,
WebKitErrorFrameLoadInterruptedByPolicyChange =             102,
WebKitErrorCannotFindPlugIn =                               200,
WebKitErrorCannotLoadPlugIn =                               201,
WebKitErrorJavaUnavailable =                                202,
WebKitErrorBlockedPlugInVersion =                           203,

Seems all web page related and we don't want to stop the auth flow because of web page error :)